flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java

```diff
@@ -481,7 +484,70 @@
 				"Map strategy fails for an odd number of arguments",
 				InputTypeStrategies.SPECIFIC_FOR_MAP)
 				.calledWithArgumentTypes(DataTypes.BIGINT(), DataTypes.BIGINT(), DataTypes.BIGINT())
-				.expectErrorMessage("Invalid number of arguments. 3 arguments passed.")
+				.expectErrorMessage("Invalid number of arguments. 3 arguments passed."),
+
+			TestSpec
+				.forStrategy(
+					"Logical type roots instead of concrete data types",
+					sequence(
+						logical(LogicalTypeRoot.VARCHAR),
+						logical(LogicalTypeRoot.DECIMAL),
+						logical(LogicalTypeRoot.DECIMAL),
+						logical(LogicalTypeRoot.BOOLEAN),
+						logical(LogicalTypeRoot.INTEGER, false)))
+				.calledWithArgumentTypes(
+					DataTypes.NULL(),
+					DataTypes.INT(),
+					DataTypes.DOUBLE(),
+					DataTypes.BOOLEAN().notNull(),
+					DataTypes.INT().notNull())
+				.expectSignature(
+					"f(<VARCHAR>, <DECIMAL>, <DECIMAL>, <BOOLEAN>, <INTEGER NOT NULL>)")
+				.expectArgumentTypes(
+					DataTypes.VARCHAR(1),
+					DataTypes.DECIMAL(10, 0),
+					DataTypes.DECIMAL(30, 15),
+					DataTypes.BOOLEAN(),
```

Comment:
How about we preserve the nullability? Right now there is no way to say that we accept both. Therefore e.g. we loose the nullability info in such case:

```
.inputTypeStrategy(logical(LogicalTypeRoot.BOOLEAN))
.outputTypeStrategy(TypeStrategies.argument(0))
```

I would suggest either:
* if the `expectedNullability` in `FamilyArgumentTypeStrategy`/`RootArgumentTypeStrategy` is false forward the nullability of the input argument
* introduce three state value -> expect nullable, not null, both